### PR TITLE
GIX-1819: Identify Swap Participation Transaction

### PR DESCRIPTION
**Pull request by lmuntaner**
# Motivation

Users can identify transactions related to participating in a swap.

# Changes

* New derived store creator "createSwapCanisterAccountsStore". Returns a derived store of AccountIdentifiers based on the passed controller and the present swap canisters.
* New parameter "swapAccounts" in "mapNnsTransaction" util.
* New parameter "swapAccounts" in "transactionType" util. It uses this new parameter to identify swap participation transactions.
* NnsTransactionCard uses the new derived store and passes it to "mapNnsTransaction" so that the swap participation can be identified.

# Tests

* New test case in NnsTransactionCard spec file.
* Test new derived store.
* Fix tests with new signature.
* Add test case in "transactionType" to check it identifies swap participation transaction.
* Add test case in "mapNnsTransaction" to check it identifies swap participation transaction.

# Todos

- [ ] Add entry to changelog (if necessary).

### CODE REVIEW

**Comment by lmuntaner** [comment]
New UI:

Screenshot 2023-08-15 at 09 02 10

**Comment by lmuntaner** [comment]
@dskloetd @peterpeterparker please review

File: frontend/src/lib/derived/sns-swap-canisters-accounts.derived.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import { getSwapCanisterAccount } from "$lib/utils/sns.utils";
+import type { AccountIdentifier } from "@dfinity/nns";
+import { Principal } from "@dfinity/principal";
+import { isNullish } from "@dfinity/utils";
+import { derived } from "svelte/store";
+
+export const createSwapCanisterAccountsStore = (controller?: Principal) =>
+  derived<SnsAggregatorStore, AccountIdentifier[]>(
+    snsAggregatorStore,
+    ($snsAggregatorStore) =>
```
Comment (dskloetd) [inline frontend/src/lib/derived/sns-swap-canisters-accounts.derived.ts:14]: Since this store will be used to check presence of certain accounts, it would be most convenient if it was of type `Set<string>` instead of `AccountIdentifier[]`.
Do you think the convenience is worth being a bit less specific about the types?
I think it would make usage more readable.

Otherwise, a convenience method would be nice so we don't need to convert it before using.

File: frontend/src/tests/lib/components/accounts/NnsTransactionCard.spec.ts [hunk 2]
```diff
@@ -43,6 +48,38 @@ describe("NnsTransactionCard", () => {
...
+          to: swapCanisterAccount.toHex(),
+        },
+      },
+    };
+    const { queryByTestId } = renderTransactionCard(
+      mockMainAccount,
+      swapTransaction
+    );
+
+    expect(queryByTestId("headline").textContent).toBe("Decentralized Swap");
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/accounts/NnsTransactionCard.spec.ts:80]: Shouldn't it be "Decentralization Swap" instead of "Decentralized Swap"?
It's a swap that decentralizes a project; not a swap that is itself decentralized, right?

I see the message was pre-existing so it can be changed in another PR. But I do think the existing message is wrong.

File: frontend/src/tests/lib/components/accounts/NnsTransactionCard.spec.ts [hunk 3]
```diff
@@ -59,7 +96,10 @@ describe("NnsTransactionCard", () => {
     const account = mockMainAccount;
     const transaction = mockSentToSubAccountTransaction;
     const { getByTestId } = renderTransactionCard(account, transaction);
-    const { displayAmount } = mapNnsTransaction({ account, transaction });
+    const { displayAmount } = mapNnsTransaction({
+      account,
+      transaction,
+    });
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/accounts/NnsTransactionCard.spec.ts:102]: Was this changed automatically or did you change it?
It looks like a comma was added which changed the formatting.
I wouldn't change it unnecessarily.

File: frontend/src/tests/lib/derived/sns-swap-canister-accounts.derived.spec.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  });
+
+  it("should convert swap canisters to accounts for a given controller", () => {
+    snsAggregatorStore.setData([aggregatorData]);
+
+    const controller = mockPrincipal;
+    const store = createSwapCanisterAccountsStore(controller);
+
+    expect(get(store)).toEqual([
+      getSwapCanisterAccount({ controller, swapCanisterId }),
```
Comment (dskloetd) [inline frontend/src/tests/lib/derived/sns-swap-canister-accounts.derived.spec.ts:30]: Can we do a more explicit comparison instead of calling the same function as in the implementation?
Preferably set an explicit value in the setup and compare to that value.

File: frontend/src/lib/derived/sns-swap-canisters-accounts.derived.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  snsAggregatorStore,
+  type SnsAggregatorStore,
+} from "$lib/stores/sns-aggregator.store";
+import { getSwapCanisterAccount } from "$lib/utils/sns.utils";
+import type { AccountIdentifier } from "@dfinity/nns";
+import { Principal } from "@dfinity/principal";
+import { isNullish } from "@dfinity/utils";
+import { derived } from "svelte/store";
+
+export const createSwapCanisterAccountsStore = (controller?: Principal) =>
```
[the remark this answers is not in the document]
Comment (lmuntaner) [reply frontend/src/lib/derived/sns-swap-canisters-accounts.derived.ts:11]: Sure

File: frontend/src/lib/derived/sns-swap-canisters-accounts.derived.ts (hunk 1 above)
Comment (lmuntaner) [reply frontend/src/lib/derived/sns-swap-canisters-accounts.derived.ts:14]: Good point. Using a set will be good.

File: frontend/src/tests/lib/components/accounts/NnsTransactionCard.spec.ts (hunk 2 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/components/accounts/NnsTransactionCard.spec.ts:80]: I agree. Let's change it in this PR.

File: frontend/src/tests/lib/components/accounts/NnsTransactionCard.spec.ts (hunk 3 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/components/accounts/NnsTransactionCard.spec.ts:102]: I'll revert.

File: frontend/src/tests/lib/derived/sns-swap-canister-accounts.derived.spec.ts (hunk 4 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/derived/sns-swap-canister-accounts.derived.spec.ts:30]: Not exactly.

The account identifier isn't in the setup. And from an account identifier I can't tell the controller. So I can't start with an account identifier and get the swap canister.

Instead, I can compare to a string. But I'm not sure this will help much.

**Comment by lmuntaner** [comment]
@dskloetd ready for another review

File: frontend/src/lib/derived/sns-swap-canisters-accounts.derived.ts [hunk 6]
```diff
@@ -3,21 +3,29 @@ import {
...
     snsAggregatorStore,
     ($snsAggregatorStore) =>
       isNullish(controller) || isNullish($snsAggregatorStore.data)
-        ? []
-        : $snsAggregatorStore.data.map(({ canister_ids }) =>
-            getSwapCanisterAccount({
-              controller,
-              swapCanisterId: Principal.fromText(canister_ids.swap_canister_id),
-            })
+        ? (new Set() as Set<string>)
```
[the remark this answers is not in the document]
Comment (lmuntaner) [reply frontend/src/lib/derived/sns-swap-canisters-accounts.derived.ts]: done

File: frontend/src/lib/derived/sns-swap-canisters-accounts.derived.ts [hunk 7]
```diff
@@ -16,16 +16,18 @@ export const createSwapCanisterAccountsStore = (controller?: Principal) =>
...
-              getSwapCanisterAccount({
-                controller,
-                swapCanisterId: Principal.fromText(
-                  canister_ids.swap_canister_id
-                ),
-              }).toHex()
+      new Set(
+        isNullish(controller) || isNullish($snsAggregatorStore.data)
+          ? undefined
+          : new Set(
```
Comment (dskloetd) [inline frontend/src/lib/derived/sns-swap-canisters-accounts.derived.ts:22]: You don't need this `new Set(` here now that you have it on line 19.

Comment (lmuntaner) [reply frontend/src/lib/derived/sns-swap-canisters-accounts.derived.ts:22]: oh, no yes!

Comment (lmuntaner) [reply frontend/src/lib/derived/sns-swap-canisters-accounts.derived.ts:22]: https://github.com/dfinity/nns-dapp/pull/3115
